stm/stm.py

```diff
@@ -0,0 +1,418 @@
+        """
+        Select a subset of the STM
+
+        Parameters
+        ----------
+        method : str
+            Method of subsetting. Choose from "threshold", "density" and "polygon".
+            - threshold: select all points with a threshold criterion, e.g.
+                data_xr.stm.subset(method="threshold", var="thres", threshold='>1')
+            - density: select one point in every [dx, dy] cell, e.g.
+                data_xr.stm.subset(method='density', dx=0.1, dy=0.1)
+            - polygon: select all points inside a given polygon, e.g.
+                data_xr.stm.subset(method='polygon', polygon=path_polygon_file)
+                or
+                import geopandas as gpd
+                polygon = gpd.read_file(path_polygon_file)
+                data_xr.stm.subset(method='polygon', polygon=polygon)
+
+        Returns
+        -------
+        xarray.Dataset
+            A subset of the original STM.
+        """
+
+        match method:  # Match statements available only from python 3.10 onwards
+            case "threshold":
+                _check_threshold_kwargs(
+                    **kwargs
+                )  # Check is all required kwargs are available
+                if kwargs["threshold"][0] == "<":
+                    str_parts = kwargs["threshold"].partition("<")
+                    check_mult_relops(
+                        str_parts[2]
+                    )  # Check to ensure multiple relational operators are not present
+                    data_xr_subset = self._obj.where(
+                        self._obj[kwargs["var"]] < float(str_parts[2]), drop=True
+                    )
+                elif kwargs["threshold"][0] == ">":
+                    str_parts = kwargs["threshold"].partition(">")
+                    check_mult_relops(str_parts[2])
+                    data_xr_subset = self._obj.where(
+                        self._obj[kwargs["var"]] > float(str_parts[2]), drop=True
+                    )
+                else:
+                    raise Exception(
+                        "Suitable relational operator not found! Please check input"
+                    )
+            case "density":
+                check_density_kwargs(**kwargs)  # Check for all require kwargs
+                gdf = gpd.GeoDataFrame(
+                    self._obj["points"],
+                    geometry=gpd.points_from_xy(
+                        self._obj[kwargs["x"]], self._obj[kwargs["y"]]
+                    ),
+                )
+                grid_cell = ((shapes) for shapes in zip(gdf.geometry, gdf.index))
+                out_x = math.ceil(
+                    (max(self._obj[kwargs["x"]]) - min(self._obj[kwargs["x"]]))
+                    / kwargs["dx"]
+                )
+                out_y = math.ceil(
+                    (max(self._obj[kwargs["y"]]) - min(self._obj[kwargs["y"]]))
+                    / kwargs["dy"]
+                )
+                raster = features.rasterize(
```

Comment:
A possible draw back is having to create ever finer rasters for higher desired densities, as well as the need for possibly non-intuuitive cell sizes.

E.g: select at density 1/m^2 -> grid of dx = dy = 1m
how to select at e.g 5/m^2

Reply:
Thx for the heads up Meiert! For this function, the application is only for select points to make a network, and there should not be too short arcs within the network. Then I think the requirement is to ensure there is no second point within a specific range. Therefore I think the current implementation is fine.